File: Data Structures & Algorithms/generate-parentheses/submission-1.cpp
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        string s = "";
        vector<string> res;
        backtracking(n, 0, 0, s, res);
        return res;
    }

private:
    void backtracking(int n, int o, int c, string& s, vector<string>& res) {
        if(o == n and c == n) {
            res.push_back(s);
            return;
        }
        
        if(o < n) {
            s.push_back('(');
            backtracking(n, o + 1, c, s, res);
            s.pop_back();
        }

        if(c < o) {
            s.push_back(')');
            backtracking(n, o, c + 1, s, res);
            s.pop_back();
        }
    }
};

File: Data Structures & Algorithms/combination-target-sum-ii/submission-1.py
class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        candidates.sort()
        res = []
        curr = []

        def backtracking(i, target):
            if target == 0:
                res.append(curr[:])
                return
            if target < 0 or i >= len(candidates):
                return
            
            curr.append(candidates[i])
            backtracking(i + 1, target - candidates[i])
            curr.pop()
            j = i + 1
            while j < len(candidates) and candidates[j] == candidates[i]:
                j += 1
            backtracking(j, target)
        
        backtracking(0, target)
        return res